[sw/device] Remove newlib

alistair23 opened the pull request:
We currently include newlib in OT for the `assert()`, `memcpy()`, `memset()` and `strlen()` functions. This has the following disadvantages.

1. License issues. newlib is a confusing mixture of licenses including GPL (which is not compatible with Apache 2.0). This adds a burden to developers having to understand licensing issues when using newlib. Due to the complex newlib license I suspect this results with lawyers being required to decide if and when we can statically link with it.
2. Code size. With newlib my `build-fpga/sw/device/examples/hello_world/hello_world.bin` binary is 6040 bytes. After this series it drops to 5688. That's about a 6% size reduction.
3. Build issues. Requiring users to have newlib cross compiled for RISC-V ilp32 is just another burden in the cross compile process.
4. Security, including newlib in boot ROM increases the effort required to audit the entire code base.

This series removes the requirement of newlib for sw/device binaries.

It does this by re-implemened `abort()`, `memcpy()`, `memset()` and `strlen()`. There are basic implementations, if desired they can be optimised in the future (for ROM it doesn't seem too important and simplicity seems more important).

The current only assert() is assert(1), so replacing it with abort() doesn't make any real changes.

I have removed a large number of current #includes to simplify our dependencies. I haven't removed all of them, that can be left for another day.

alistair23 commented:
This fails the linter as I purposely didn't change the newlib code. Let me know if you want me to change it to match OT style.

tjaychen commented:
i think others can probably better comment if we should be copying over the newlib code..but assume for now we do, the right way we're supposed to do it include it under sw/vendor.

The lint checks also avoid the sw/vendor directory I believe.

imphil commented:
Thanks for starting this discussion @alistair23! 

Copying in parts of a C library into our project will make it harder to develop portable code for OpenTitan. While I can see the motivation for this change, I'd like to ensure that we have a full understanding of the pros and cons here before making a move. I'd also like to discuss alternatives, like using bionic. Can you add this topic to the agenda of the next software meeting?


My take on the points you brought up:

> License issues. newlib is a confusing mixture of licenses including GPL (which is not compatible with Apache 2.0). This adds a burden to developers having to understand licensing issues when using newlib. Due to the complex newlib license I suspect this results with lawyers being required to decide if and when we can statically link with it.

Yes, the licensing situation of newlib is confusing, but not too bad: the relevant parts of newlib are all BSD-like licensed. We should be able to figure out guidelines or technical measures that allow use to use newlib without including GPL-licensed code (and I wouldn't be surprised if that exercise hasn't been done by others before, given how widely newlib is used). 

Vendoring in the non-GPL parts of newlib is one option, but I can think of other options as well: remove GPL parts of newlib when building the toolchain, using another C library, or a lint-style scanner that flags the use GPL-licensed functionality.

> Code size. With newlib my build-fpga/sw/device/examples/hello_world/hello_world.bin binary is 6040 bytes. After this series it drops to 5688. That's about a 6% size reduction.

We should be able to achieve similar binary size reductions by eliminating more dead code in the compilation and linking process. Have you played around with the settings there?

> Build issues. Requiring users to have newlib cross compiled for RISC-V ilp32 is just another burden in the cross compile process.

We do provide pre-built toolchains which can be reproduced easily by others (https://github.com/lowRISC/lowrisc-toolchains). With the desire of moving to clang/LLVM the toolchain build will  become more involved anyways, which reduces the win we get from not building newlib.

gkelly commented:
> I'd also like to discuss alternatives, like using bionic. Can you add this topic to the agenda of the next software meeting?

I agree, I think especially before we start forking libraries we should consider how we can gracefully continue to use them, or failing that considering the alternatives.  I've added this to the software meeting agenda.

asb commented:
There's also [picolibc](https://github.com/keith-packard/picolibc). I think the GPL vs non-GPL portions of newlib are fairly well understood in general, but if you have a specific concern there it's probably best to get in touch with legal-committee@lowrisc.org.

I was really interested in the work done through ellcc on supporting a bare-metal musl, but work there seems to have died down. I know some companies have their own internal bare-metal musl ports too. More recently, there's picolibc of course.

alistair23 commented:
There are a few things to consider here.

Really the main one is why do you need a libc? What is the boot ROM and low level software doing that needs a libc? I don't see a good justification for including one.

After this series there are three functions from a libc included. Even then the `strlen()` function can probably be removed, resulting in just a `memcpy()` and `memset()`.

As for using  bionic that seems like overkill. picolibc is small, but again why do we need a libc? Also using random libcs in a security focused context, focused on reproducible builds seems unnecessary when it can be avoided. What happens when newlib, bionic or picolib have a bug in the version used in the OT ROM?

There also seems to be this assumption that OT can only be built with the one approved toolchain. This is not going to be the case forever, developers will expect to use their host toolchain to build software. When I build any other project (for RISC-V or x86) I use my Linux distros toolchain instead of installing that project's special one. As RISC-V matures this is going to be more and more common. Expecting everyone to use a specific toolchain (not from upstream) seems unreasonable.

As people use their own toolchains you won't be able to remove GPL parts of newlib. Even if you are removing the GPL parts, then you have a fork, at which point why not just move the code into your mono repo?

moidx commented:
> Really the main one is why do you need a libc? What is the boot ROM and low level software doing that needs a libc? I don't see a good justification for including one.

I don't think we are using libc in boot ROM. If this is the case, we should file an issue and get it fixed. Once we start working on a production ROM, we should consider isolating the ROM libraries to ensure this requirement is easier to verify.

> There also seems to be this assumption that OT can only be built with the one approved toolchain. This is not going to be the case forever, developers will expect to use their host toolchain to build software. When I build any other project (for RISC-V or x86) I use my Linux distros toolchain instead of installing that project's special one. As RISC-V matures this is going to be more and more common. Expecting everyone to use a specific toolchain (not from upstream) seems unreasonable.

We can certainly add support for multiple toolchains in the future, but having a reproducible build environment is still a requirement.  For some of our DV use cases, we'll support compiling the libraries as x86 binaries for integration with simulation test benches.

alistair23 commented:
> I don't think we are using libc in boot ROM. If this is the case, we should file an issue and get it fixed. Once we start working on a production ROM, we should consider isolating the ROM libraries to ensure this requirement is easier to verify.

It is included, this series include a patch to remove it. https://github.com/lowRISC/opentitan/pull/900/commits/fd6c7c1d3d5e37e96cc23433ef501ad933bd0b81

> We can certainly add support for multiple toolchains in the future, but having a reproducible build environment is still a requirement. For some of our DV use cases, we'll support compiling the libraries as x86 binaries for integration with simulation test benches.

You don't need to support multiple toolchains, just don't expect everyone to be using the same one you are using. As for reproducible builds that is another problem, but it doesn't mean that only your toolchain and libc work, which would be the case if you use bionic (rarely used outside of Google) or a non-GPL fork of newlib as mentioned earlier.

tjaychen commented:
since it seems like there's general agreement that rom should not reference libs, but more discussion is needed on exactly which lib (if any) to use, should the rom related change be split into a different PR? That way it won't have to linger.

alistair23 commented:
I would prefer that they both go in together. I think it's confusing removing it from one and not the other.

asb commented:
I think we should really separate out the issues in play:
* Should boot ROM require libc
  * The consensus is no
* Should the demo applications require libc
  * No consensus yet
  * Also note - these really are "demo" applications so we should be careful of spending too much time arguing for their requirements
* What libc should the project use when libc is required
* Should we copy in selected libc functions and if so, from what source?
  * If we do go down that route and it's a really trivial number of simple functions, I'd rather we just write them or import from a source that has the simplest possible licensing situation (e.g. uniform BSD/MIT/Apache).

Alistair - could you elaborate on why you think the decision for boot ROM and later stage applications needs to be linked? They seemed like separable questions to me, but I could be missing something.

To respond to the point about toolchains: I totally agree that the OpenTitan project shouldn't put specialised requirements on the toolchain. We are going to have a specific toolchain for official builds and ease of getting started, but other toolchains should work. That said, I don't think that asking users to have an ELF toolchain with newlib is onerous or unexpected for an embedded project.

alistair23 commented:
> * Should boot ROM require libc

I'm glad everyone agrees with this.

> * Should the demo applications require libc

It isn't all the demo applications. I haven't changed coremark or any of the larger applications, just the basic libs.

Note that: https://github.com/lowRISC/opentitan/blob/master/sw/device/exts/common/_crt.c requires libc which is commonly used by the libs and the boot ROM.

> * Should we copy in selected libc functions and if so, from what source?

I don't mind where it comes from, I went with newlib as that is what you are already using.

> Alistair - could you please elaborate on why you think the decision for boot ROM and later stage applications needs to be linked? They seemed like separable questions to me, but I could be missing something.

The boot ROM has a dependency on https://github.com/lowRISC/opentitan/blob/master/sw/device/exts/common/_crt.c which depends on libc. It would be a pain to remove the libc requirement from the boot ROM but not the rest of the libcs. It's not impossible, but it seems like unnecessary complexity.

> To respond to the point about toolchains: I totally agree that the OpenTitan project shouldn't put specialised requirements on the toolchain. We are going to have a specific toolchain for official builds and ease of getting started, but other toolchains should work. That said, I don't think that asking users to have an ELF toolchain with newlib is onerous or unexpected for an embedded project.

Great! Some of the things I have seen led me to worry that it was expected that everyone uses a special lowRISC toolchain. I'm all for having something to help people get started, but it shouldn't be a requirement.

I agree that asking people to have newlib isn't crazy, I just don't see the justification of needing a libc for ROM and small baremetal demos.

alistair23 commented:
Ping? Any more thoughts on removing libc from the ROM?

gkelly commented:
Yes, we covered this in the software meeting today.  The consensus was that yes: we will remove newlib from the build for bootrom code, but will do so by reimplementing the parts of `stddef.h` and `strings.h` that we need and not by including the newlib sources directly, which has a higher burden than reimplementation.

alistair23 commented:
So with the three functions re-written this PR is good to go?

alistair23 commented:
Also, why re-write the three functions? We already have an implementation that has been reviewed by lots of people, it seems unnecessary to re-write it from scratch.

alistair23 commented:
To avoid newlib license issues we could use the musl implementation which has also had a lot of eyes on it.

alistair23 commented:
I have pushed an update that re-writes the functions.

alistair23 commented:
I have added some more commits, here is where it stands:
 - I have re-implemened `memcpy()`, `memset()` and `strlen()`. There are basic implementations, if desired they can be optimised in the future (for ROM it doesn't seem too important and simplicity seems more important.
   - I have avoided using `size_t` to not require any new `#includes`
 - I have removed a large number of current `#includes` to simplify our dependencies. I haven't removed all of them, that can be left for another day.

This addresses all concerns about using differently licensed implementations of libc (as it's now just written internally) that have been discussed. Let me know if there are any other comments.

alistair23 commented:
`build-fpga/sw/device/boot_rom/boot_rom.elf` went from `86776` bytes to `80296` with this change.

silvestrst commented on sw/device/lib/gpio.c:
```diff
@@ -20,7 +18,8 @@ void gpio_write_bit(unsigned int bit, unsigned int val) {
     gpio_masked_out_reg = (uint32_t *)GPIO_MASKED_OUT_UPPER(0);
   } else {
-    assert(1 && "bit must be < 32");
+    uart_send_str("bit must be less then 32\n");
```
Using uart_send_str isntead of assert changes the behaviour - as it won't abort the execution. As the function is void, there is no way to convey this information to the caller. Is it worth makin this function return bool?

alistair23 commented on sw/device/lib/common.h:
```diff
@@ -41,4 +38,9 @@ static const unsigned long UART_BAUD_RATE = 230400;
 #define BITLENGTH(X) \
   ((BITLENGTH_5(BITLENGTH_4(BITLENGTH_3(BITLENGTH_2(BITLENGTH_1(X)))))) & 0x7f)
 
+void *memcpy(void *dest, const void *src, int len);
```
Yep, seems fair.

I originally had it like that but wanted to remove the `stddef.h` include. I have changed these back to match the standard and now include `stddef.h`.

alistair23 commented:
I have updated the PR description and the patches to address everyone's comments. Let me know if there are any other issues.

lenary commented on sw/device/lib/gpio.c:
```diff
@@ -20,7 +20,7 @@ void gpio_write_bit(unsigned int bit, unsigned int val) {
   } else if (bit < 32) {
     gpio_masked_out_reg = (uint32_t *)GPIO_MASKED_OUT_UPPER(0);
   } else {
-    assert(1 && "bit must be < 32");
```
It's useful to not have a dependency between GPIO and uart.

I think a reasonable thing to do is include the old message as a comment before the call to abort, so you can see it when you debug, but so the dependency is avoided.

silvestrst replied:
Fair point, agreed!

alistair23 commented:
I have updated the PR

alistair23 commented on sw/device/lib/memcpy.c:
```diff
@@ -0,0 +1,16 @@
+
+#include "sw/device/lib/common.h"
+
+void *memcpy(void *dest, const void *src, size_t n) {
```
Sorry about that. I have added it now.

silvestrst commented:
I think you have missed my review comment above:
"sw/device/lib/gpio.c still has <assert.h> include, which is now redundant and should be removed"

With this little change, I don't have any more comments.

Thank you for updating your patches.

alistair23 commented:
> I think you have missed my review comment above:
> "sw/device/lib/gpio.c still has <assert.h> include, which is now redundant and should be removed"

Ah I missed that. Good catch. I have removed it now

silvestrst commented:
Thank you Alistair, looks good to me, I don't have any more comments!
